core/test/dataset_next_test.cpp

```diff
@@ -0,0 +1,857 @@
+
+  const auto reference =
+      makeVariable<double>({{Dim::X, 3}, {Dim::Y, 2}}, {1, 2, 0, 0, 5, 6});
+  EXPECT_EQ(d.coords()[Dim::X], reference);
+}
+
+TEST(CoordsConstProxy, slice_bin_edges_with_2D_coord) {
+  next::Dataset d;
+  const auto x = makeVariable<double>({{Dim::Y, 2}, {Dim::X, 2}}, {1, 2, 3, 4});
+  const auto y_edges = makeVariable<double>({Dim::Y, 3}, {1, 2, 3});
+  d.setCoord(Dim::X, x);
+  d.setCoord(Dim::Y, y_edges);
+  const auto coords = d.coords();
+
+  const auto sliceX = coords.slice({Dim::X, 1});
+  EXPECT_ANY_THROW(sliceX[Dim::X]);
+  EXPECT_EQ(sliceX[Dim::Y], coords[Dim::Y]);
+
+  const auto sliceDX = coords.slice({Dim::X, 1, 2});
+  EXPECT_EQ(sliceDX[Dim::X].dimensions(),
+            Dimensions({{Dim::Y, 2}, {Dim::X, 1}}));
+  EXPECT_EQ(sliceDX[Dim::Y], coords[Dim::Y]);
+
+  const auto sliceY = coords.slice({Dim::Y, 1});
+  EXPECT_ANY_THROW(sliceY[Dim::X]);
+  EXPECT_ANY_THROW(sliceY[Dim::Y]);
+
+  const auto sliceY_edge = coords.slice({Dim::Y, 1, 2});
+  // The X coord is dropped if a single bin is in the range.
+  EXPECT_ANY_THROW(sliceY_edge[Dim::X]);
+  EXPECT_EQ(sliceY_edge[Dim::Y].dimensions(), Dimensions({Dim::Y, 1}));
+
+  const auto sliceY_bin = coords.slice({Dim::Y, 1, 3});
+  EXPECT_EQ(sliceY_bin[Dim::X].dimensions(),
+            Dimensions({{Dim::Y, 1}, {Dim::X, 2}}));
+  EXPECT_EQ(sliceY_bin[Dim::Y].dimensions(), Dimensions({Dim::Y, 2}));
+}
+
+// Using typed tests for common functionality of DataProxy and DataConstProxy.
+template <typename T> class DataProxyTest : public ::testing::Test {
+public:
+  using proxy_type = T;
+};
+
+using DataProxyTypes = ::testing::Types<next::Dataset, const next::Dataset>;
+TYPED_TEST_CASE(DataProxyTest, DataProxyTypes);
+
+TYPED_TEST(DataProxyTest, isSparse_sparseDim) {
+  next::Dataset d;
+  typename TestFixture::proxy_type &d_ref(d);
+
+  d.setValues("dense", makeVariable<double>({}));
+  ASSERT_FALSE(d_ref["dense"].isSparse());
+  ASSERT_EQ(d_ref["dense"].sparseDim(), Dim::Invalid);
+
+  d.setValues("sparse_data", makeSparseVariable<double>({}, Dim::X));
+  ASSERT_TRUE(d_ref["sparse_data"].isSparse());
+  ASSERT_EQ(d_ref["sparse_data"].sparseDim(), Dim::X);
+
+  d.setSparseCoord("sparse_coord", makeSparseVariable<double>({}, Dim::X));
+  ASSERT_TRUE(d_ref["sparse_coord"].isSparse());
+  ASSERT_EQ(d_ref["sparse_coord"].sparseDim(), Dim::X);
+}
+
+TYPED_TEST(DataProxyTest, dims) {
+  next::Dataset d;
+  const auto dense = makeVariable<double>({{Dim::X, 1}, {Dim::Y, 2}});
+  const auto sparse =
+      makeSparseVariable<double>({{Dim::X, 1}, {Dim::Y, 2}}, Dim::Z);
+  typename TestFixture::proxy_type &d_ref(d);
+
+  d.setValues("dense", dense);
+  ASSERT_EQ(d_ref["dense"].dims(), dense.dimensions());
+
+  // Sparse dimension is currently not included in dims(). It is unclear whether
+  // this is the right choice. An unfinished idea involves returning
+  // std::tuple<std::span<const Dim>, std::optional<Dim>> instead, using `auto [
+  // dims, sparse ] = data.dims();`.
```

Comment:
It does seems strange to me that you call dimensions and get only get your dense dimensions, you might not know about sparse dimensions, or even know of the separate API aspects for them. I think the design principle of "api's should be hard to use incorrectly" could be applied here. The `std::optional` approach seems like a good one even though there is more work for the client when they are tending to use dense dimensions.

Reply:
I also like that principle. But on the other hand in this case it would force 100% of the users to learn about sparse dimensions, while only 10% may actually need it. Simple things should be simple to do, and returning a pair/struct here is certainly not simple.

Another solution would be to move the sparse dim inside the `Dimensions` class. Then this question becomes one of the interface of `class Dimensions`.
- What should `Dimensions::labels()` return?
- What should `Dimensions::shape()` return?

Reply:
One option would be to return all labels (including the sparse one) from `Dimensions::labels()`, but only the extents of the dense dimensions in `Dimensions::shape()` (and by conjecture, `Dimensions::operator[]` would throw if provided with the sparse dimension label). This would:

- Provide convenient access to all labels.
- Imply no changes for anyone not using sparse dimensions.
- Iteration would fail on incorrect use (as desired):
  ```cpp
  // Natural code that works for dense data, but throws if it is not
  for (const auto dim : dims.labels())
    volume *= dims[dim];
   ```
- Code that wants to handle the sparse dimension can do so:
  ```cpp
  for (const auto dim : dims.labels())
    if (dim != dims.sparseDim())
      volume *= dims[dim];
  ```

Reply:
See https://github.com/scipp/scipp/issues/238.